### PR TITLE
[Rust] Port the ListView demo

### PR DESCRIPTION
**Pull request by ondrejkolin**
I am not sure if I should put those comments into the code...

### CODE REVIEW

File: src/List View/code.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    // Required by gtk::StringList::new(), otherwise crashes
+    gtk::init().unwrap();
+
+    let list_view: gtk::ListView = workbench::builder().object("list_view").unwrap();
+    let add: gtk::Button = workbench::builder().object("add").unwrap();
+    let remove: gtk::Button = workbench::builder().object("remove").unwrap();
+
+    // https://doc.rust-lang.org/std/cell/struct.Cell.html
+    // "A mutable memory location."
+    // Makes possible accessing and changing values from signal handlers
```
Comment (Hofer-Julian) [inline src/List View/code.rs:17]:
```suggestion
    // Makes it possible to access and change values from within signal handlers
```

File: src/List View/code.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            model
+                .item(selected_item) // Get the item
+                .unwrap() // Make sure it exists
+                .downcast::<gtk::StringObject>() // It's a member of GStringList
+                .unwrap() // Make sure it's really a StringObject
+                .string() // Read the string value
+        )
+    });
+
+    add.connect_clicked(gtk::glib::clone!(
```
Comment (Hofer-Julian) [inline src/List View/code.rs:48]:
```suggestion
    add.connect_clicked(glib::clone!(
```

File: src/List View/code.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        // Get the item counter value
+        let value = item.get();
+        // Access the underlying gtk::StringList
+        let strings_model: gtk::StringList = model
+            .model()
+            .unwrap()
+            .downcast::<gtk::StringList>()
+            .unwrap();
+        strings_model.append(format!("New item {}", item.get()).as_str());
+        //Increase the counter
```
Comment (Hofer-Julian) [inline src/List View/code.rs:61]:
```suggestion
        // Increase the counter
```

File: src/List View/code.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        item.set(value + 1);
+    }));
+
+    remove.connect_clicked(gtk::glib::clone!(
+    // Copy the reference, so it's accessible from the closure
+    // https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/macro.clone.html
+    @weak model => move |_| {
+      let selected_item = model.selected();
+      // In order to delete values we need to access the
+      // actuall Stringlist model, we've created.
```
Comment (Hofer-Julian) [inline src/List View/code.rs:71]:
```suggestion
      // actual StringList model, we've created.
```

**Review by Hofer-Julian — COMMENTED** [review]
A few minor comments.

Your comments are fine by me

File: src/List View/code.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    // Model
+    let strings_model =
+        gtk::StringList::new(&["Default Item 1", "Default Item 2", "Default Item 3"]);
+    let model = gtk::SingleSelection::new(Some(strings_model));
+
+    //View
+    model.connect_items_changed(move |_model, position, removed, added| {
+        println!(
+            "position: {}, Item removed? {} Item added? {}",
```
Comment (Hofer-Julian) [inline src/List View/code.rs:28]:
```suggestion
            "position: {}, Item removed? {}, Item added? {}",
```

File: src/List View/code.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            position,
+            removed > 0,
+            added > 0
+        )
+    });
+
+    model.connect_selection_changed(move |model, _position, _n_items| {
+        let selected_item = model.selected();
+        println!(
+            "Model item selected from view: {:?}",
```
Comment (Hofer-Julian) [inline src/List View/code.rs:38]:
```suggestion
            "Model item selected from view: {}",
```

File: src/List View/code.rs (hunk 5 above)
Comment (ondrejkolin) [reply src/List View/code.rs:28]: @Hofer-Julian This one produces quite interesting values, it differs from other languages. Is that expected based on your gtk-rs knowledge?

**Comment by ondrejkolin** [comment]
@ondrejkolin btw. the `position: {}, Item removed? {}, Item added? {}` differs from the other languages... Is that because the model is immutable in rust?

File: src/List View/code.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    });
+
+    add.connect_clicked(glib::clone!(
+    // Copy the reference, so it's accessible from the closure
+        // https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/macro.clone.html
+    @weak model => move |_| {
+        // Get the item counter value
+        let value = item.get();
+        // Access the underlying gtk::StringList
+        let strings_model: gtk::StringList = model
```
Comment (Hofer-Julian) [inline src/List View/code.rs:55]:
```suggestion
        let strings_model = model
```

File: src/List View/code.rs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    }));
+
+    remove.connect_clicked(gtk::glib::clone!(
+    // Copy the reference, so it's accessible from the closure
+    // https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/macro.clone.html
+    @weak model => move |_| {
+      let selected_item = model.selected();
+      // In order to delete values we need to access the
+      // actual StringList model, we've created.
+      let strings_model: gtk::StringList = model
```
Comment (Hofer-Julian) [inline src/List View/code.rs:72]:
```suggestion
      let strings_model = model
```

File: src/List View/code.rs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    // Makes it possible to access and change values from within signal handlers
+    let item = Cell::new(1);
+
+    // Model
+    let strings_model =
+        gtk::StringList::new(&["Default Item 1", "Default Item 2", "Default Item 3"]);
+    let model = gtk::SingleSelection::new(Some(strings_model));
+
+    //View
+    model.connect_items_changed(move |_model, position, removed, added| {
```
Comment (Hofer-Julian) [inline src/List View/code.rs:26]:
```suggestion
    let model = gtk::SingleSelection::new(Some(strings_model.clone()));

    // View
    strings_model.connect_items_changed(move |_model, position, removed, added| {
```

**Review by Hofer-Julian — COMMENTED** [review]
>  btw. the position: {}, Item removed? {}, Item added? {} differs from the other languages... Is that because the model is immutable in rust?


Good that you mention that, there was indeed a subtle bug :D

Also, please rename `strings_model` to `string_model`

**Review by Hofer-Julian — APPROVED** [review]

**Pull request merged**